src/coreclr/jit/forwardsub.cpp

```diff
@@ -667,10 +661,23 @@ bool Compiler::fgForwardSubStatement(Statement* stmt)
     //
     // We may sometimes lose or change a type handle. Avoid substituting if so.
     //
-    if (gtGetStructHandleIfPresent(fwdSubNode) != gtGetStructHandleIfPresent(fsv.GetNode()))
+    // However, we allow free substitution of hardware SIMD types.
+    //
+    CORINFO_CLASS_HANDLE fwdHnd = gtGetStructHandleIfPresent(fwdSubNode);
+    CORINFO_CLASS_HANDLE useHnd = gtGetStructHandleIfPresent(fsv.GetNode());
+    if (fwdHnd != useHnd)
     {
-        JITDUMP(" would change struct handle (substitution)\n");
-        return false;
+        if ((fwdHnd == NO_CLASS_HANDLE) || (useHnd == NO_CLASS_HANDLE))
+        {
+            JITDUMP(" would add/remove struct handle (substitution)\n");
+            return false;
+        }
+
+        if (!isHWSIMDClass(fwdHnd) || !isHWSIMDClass(useHnd))
```

Comment:
What are the concerns around this for `Vector<T>` and `Vector2/3/4` (which are the non-HW `SIMD` handle types)

Reply:
Nothing particular. We can do this for all SIMD types if that makes sense.

Reply:
This would not be correct because of the aforementioned call arguments issue.

Case of interest:
```cs
Vector128<int> a = Vector4.Zero;
Call(a);
```
(Note that the particular example would be ok, as we would wrap `a` in an `OBJ(ADDR)`, but there are some cases with temps where we don't do that)

Reply:
@SingleAccretion can you build a repro for this using:
```csharp
Vector128<int> a = Vector4.Zero.AsVector128();
Call(a);
```

This will, today, have the `Vector4` reinterpreted as a `Vector128` in import (the `.AsVector128` is just dropped). So I would expect this to already be an issue...

Reply:
It is currently not an issue because `impNormStructVal` makes sure to either wrap things with an `OBJ<correct handle>`, or spill to a local of the correct type.

The sequence of events where things would "go wrong" would be along the lines of:
1) We spill `Vector4.Zero` to a temp typed as `Vector128<int>`.
2) We forward-substitute into said temp. (which cannot happen today)

It is not a theoretical problem. I've hit a case #69965 that was similar in spirit (only there the handle pre-morph was correct, but lost after morphing the argument).